pkg/utils/devlink.go

```diff
@@ -0,0 +1,267 @@
+package utils
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netlink/nl"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// DevlinkCmdInfoGet command ID
+	DevlinkCmdInfoGet = 51 // DEVLINK_CMD_INFO_GET
+	// DevlinkAttrInfoVersionFixed is nested structure for fixed values
+	DevlinkAttrInfoVersionFixed = 100 // DEVLINK_ATTR_INFO_VERSION_FIXED
+	// DevlinkAttrInfoVersionRunning is nested structure for current values
+	DevlinkAttrInfoVersionRunning = 101 // DEVLINK_ATTR_INFO_VERSION_RUNNING
+	// DevlinkAttrInfoVersionStored is nested structure for stored values
+	DevlinkAttrInfoVersionStored = 102 // DEVLINK_ATTR_INFO_VERSION_STORED
+	// DevlinkAttrInfoVersionName is value that indicates info label
+	DevlinkAttrInfoVersionName = 103 // DEVLINK_ATTR_INFO_VERSION_NAME
+	// DevlinkAttrInfoVersionValue is value that indicates info value
+	DevlinkAttrInfoVersionValue = 104 // DEVLINK_ATTR_INFO_VERSION_VALUE
+	// FwAppNameKey to extract DDP name
+	FwAppNameKey = "fw.app.name"
+
+	pciBus        = "pci"
+	headerSize    = 4
+	nestedAttrNum = 2
+)
+
+// devlinkInfoGetter is function that is responsible for getting devlink info message
+type devlinkInfoGetter func(bus, device string) ([]byte, error)
+
+// DevlinkInfoClient is a Devlink Info Client
+type DevlinkInfoClient struct {
+	infoGetter devlinkInfoGetter
+}
+
+// NewDevlinkInfoClient - returns new client
+func NewDevlinkInfoClient() DevlinkInfoClient {
+	return newDevlinkInfoClient(getDevlinkInfo)
+}
+
+func newDevlinkInfoClient(infoGetter devlinkInfoGetter) DevlinkInfoClient {
+	var dic DevlinkInfoClient
+	dic.infoGetter = infoGetter
+
+	return dic
+}
+
+// IsDevlinkSupportedByPCIDevice checks if PCI devie supports devlink info command
+func (dic DevlinkInfoClient) IsDevlinkSupportedByPCIDevice(device string) bool {
+	return dic.IsDevlinkSupportedByDevice(pciBus, device)
+}
+
+// IsDevlinkSupportedByDevice checks if device supports devlink info command
+func (dic DevlinkInfoClient) IsDevlinkSupportedByDevice(bus, device string) bool {
+	if _, err := dic.DevlinkGetDeviceInfoByNameAndKey(bus, device, FwAppNameKey); err != nil {
+		return false
+	}
+
+	return true
+}
+
+// DevlinkGetDDPProfiles returns DDP for selected device
+func (dic DevlinkInfoClient) DevlinkGetDDPProfiles(device string) (string, error) {
+	return dic.DevlinkGetDeviceInfoByNameAndKey(pciBus, device, FwAppNameKey)
+}
+
+// DevlinkGetDeviceInfoByNameAndKeys returns values for selected keys in device info
```

Comment:
should this be part of netlink package.

i.e in : https://github.com/vishvananda/netlink/blob/master/devlink_linux.go

Reply:
ah, it is (as of 19 hours ago) :)

https://github.com/vishvananda/netlink/commit/b10eb8fe5cf6ad397bcebe30a53d8607d9191b09

Reply:
In fact only general logic is in the netlink package. As we need to retrieve a specific value from the info map those functions have to be here.